### PR TITLE
Refactor/make goto analyzer more modular

### PR DESCRIPTION
**Pull request by martin-cs**
This should be completely functionality neutral.  It is intended to make it easier for other applications to make use of the abstract interpreter in the same way that goto-analyzer does and save duplicating the construction and verification code.  I strongly suggest reviewing commit by commit.

### CODE REVIEW

File: src/goto-analyzer/build_analyzer.cpp
Comment (tautschnig) [inline src/goto-analyzer/build_analyzer.cpp:152]: Should we perhaps fail in a visible way here?

File: src/goto-analyzer/build_analyzer.cpp
Comment (tautschnig) [inline src/goto-analyzer/build_analyzer.cpp:164]: As above: should this fail an invariant?

File: src/goto-analyzer/build_analyzer.h
Comment (tautschnig) [inline src/goto-analyzer/build_analyzer.h]: Nit picking: I'd use forward declarations here (`class namespacet; class optionst; class goto_modelt;`)

File: src/goto-analyzer/static_verifier.h
Comment (tautschnig) [inline src/goto-analyzer/static_verifier.h:58]: This change perhaps belongs in the prior commit?

File: src/goto-analyzer/static_verifier.h
Comment (tautschnig) [inline src/goto-analyzer/static_verifier.h:56]: Perhaps don't do the `clang-format off` in the earlier commit?

**Review by tautschnig — APPROVED** [review]
Not a CODEOWNER, but looks good to me except for the nit picks listed below.

File: src/goto-analyzer/build_analyzer.cpp
Comment (peterschrammel) [inline src/goto-analyzer/build_analyzer.cpp]: ⛏️  (multiple locations)
```suggestion
Module: Goto-Analyzer Command Line Option Processing
```

File: src/goto-analyzer/build_analyzer.h
Comment (peterschrammel) [inline src/goto-analyzer/build_analyzer.h:27]: ⛏️ Shouldn't this be documentation to the function below?

File: src/goto-analyzer/static_verifier.cpp
Comment (peterschrammel) [inline src/goto-analyzer/static_verifier.cpp]: Unstructured?

File: src/goto-analyzer/static_verifier.cpp
Comment (peterschrammel) [inline src/goto-analyzer/static_verifier.cpp:65]: Is this still needed?

File: src/goto-analyzer/static_verifier.h
Comment (peterschrammel) [inline src/goto-analyzer/static_verifier.h:50]: :bulb: Precise definitions of these entries would be great.

File: src/goto-analyzer/static_verifier.h
Comment (peterschrammel) [inline src/goto-analyzer/static_verifier.h:68]: ❓ Should these members really be mutable?

File: src/goto-analyzer/static_verifier.h
Comment (peterschrammel) [inline src/goto-analyzer/static_verifier.h:69]: 📖 Docs for each of them would be great.

**Review by peterschrammel — APPROVED** [review]

File: src/goto-analyzer/build_analyzer.cpp
Comment (martin-cs) [reply src/goto-analyzer/build_analyzer.cpp:152]: There is already quite a bit of error handling and logging on the caller side:

https://github.com/diffblue/cbmc/blob/ae4dca8d9fccf8eb78414641e0bdd1351f195f53/src/goto-analyzer/goto_analyzer_parse_options.cpp#L745

which I feel is probably sufficient?

File: src/goto-analyzer/build_analyzer.cpp
Comment (martin-cs) [reply src/goto-analyzer/build_analyzer.cpp:164]: Not as an invariant because it is user options.  `invalid_command_line_argument_exceptiont` maybe but see above for what error handling already exists.

File: src/goto-analyzer/build_analyzer.h
Comment (martin-cs) [reply src/goto-analyzer/build_analyzer.h]: Fixed.

File: src/goto-analyzer/static_verifier.h
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.h:58]: Ooops!  Fixed.

File: src/goto-analyzer/static_verifier.h
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.h:56]: Done.

File: src/goto-analyzer/build_analyzer.cpp
Comment (martin-cs) [reply src/goto-analyzer/build_analyzer.cpp]: Done.

File: src/goto-analyzer/build_analyzer.h
Comment (martin-cs) [reply src/goto-analyzer/build_analyzer.h:27]: Moved.

File: src/goto-analyzer/static_verifier.cpp
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.cpp]: fxied.

File: src/goto-analyzer/static_verifier.cpp
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.cpp:65]: I know there was at least one application that was using this format so I would like to leave it at least the year we say we will leave it.

File: src/goto-analyzer/build_analyzer.cpp
Comment (tautschnig) [reply src/goto-analyzer/build_analyzer.cpp:164]: Here and above: with that error handling already in place, perhaps even more of an indication that this should be `UNIMPLEMENTED` or `UNHANDLED_CASE`?

File: src/goto-analyzer/static_verifier.h
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.h:50]: Documented.

In doing so I noticed a potential bug here:
https://github.com/diffblue/cbmc/blob/b5d2a7283b39b721f0e5f811d43a8568167879bf/src/goto-analyzer/static_verifier.cpp#L272

As far as I can tell this code is not used anywhere in-tree.

File: src/goto-analyzer/static_verifier.h
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.h:69]: Documentation added.

File: src/goto-analyzer/static_verifier.h
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.h:68]: I mean... probably not but as `static_verifier_resultt` is really just a way to return a tuple of values I think it is better to just:

```
const static_verifier_resultt result(ai, property_location, "unused", ns);
```

and use them as `const` objects as is already done rather than write accesor functions.

**Comment by martin-cs** [comment]
@tautschnig any idea why the Xen CI run is failing?

File: src/goto-analyzer/build_analyzer.cpp
Comment (martin-cs) [reply src/goto-analyzer/build_analyzer.cpp:164]: I don't think it should be an invariant.  I think the correct thing to do is to prune the command-line options for `goto-analyzer` but that shouldn't be in a refactor commit.

**Comment by martin-cs** [comment]
It looks like

```
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/p/python2.7/libpython2.7_2.7.18-1~20.04_amd64.deb  404  Not Found [IP: 52.147.219.192 80]
```

the docker container can't be built correctly.  I will try again and then might merge if all other tests pass as this is just a refactor and doesn't touch any of the code used for Xen.

**Comment by tautschnig** [comment]
> It looks like
> 
> ```
> E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/p/python2.7/libpython2.7_2.7.18-1~20.04_amd64.deb  404  Not Found [IP: 52.147.219.192 80]
> ```
> 
> the docker container can't be built correctly. I will try again and then might merge if all other tests pass as this is just a refactor and doesn't touch any of the code used for Xen.

Will be fixed once #5926 is merged.

**Pull request merged**
